hazelcast/src/hazelcast/cp/cp.cpp

```diff
@@ -653,6 +665,257 @@ namespace hazelcast {
+                auto request = client::protocol::codec::semaphore_acquire_encode(group_id_, object_name_,
+                                                                                 internal::session::proxy_session_manager::NO_SESSION_ID,
+                                                                                 thread_id, invocation_uid, permits,
+                                                                                 timeout.count());
+                return spi::impl::ClientInvocation::create(context_, request, object_name_)->invoke().then(
+                        boost::launch::deferred, [=](boost::future<protocol::ClientMessage> f) {
+                            try {
+                                auto acquired = f.get().get_first_fixed_sized_field<bool>();
+                                if (!acquired) {
+                                    session_manager_.release_session(group_id_, session_id);
+                                }
+                                // first bool means acquired or not, second bool means if should try again
+                                return std::make_pair(acquired, false);
+                            } catch (exception::SessionExpiredException &) {
+                                session_manager_.invalidate_session(group_id_, session_id);
+                                if (use_timeout && (timeout - (std::chrono::steady_clock::now() - start) <= std::chrono::milliseconds::zero())) {
+                                    return std::make_pair(false, false);
+                                }
+                                return std::make_pair(false, true);
+                            } catch (exception::WaitKeyCancelledException &) {
+                                session_manager_.release_session(group_id_, session_id, permits);
+                                if (!use_timeout) {
+                                    BOOST_THROW_EXCEPTION(
+                                            exception::IllegalStateException("session_semaphore::try_acquire_for_millis", (boost::format(
+                                                    "Semaphore[%1%] not acquired because the acquire call on the CP group is cancelled, possibly because of another indeterminate call from the same thread.") %
+                                                                                                            object_name_).str()));
+                                }
+                                return std::make_pair(false, false);
+                            }
+                        });
+            });
+
+            return do_try_acquire_once().then(boost::launch::deferred, [=] (boost::future<std::pair<bool, bool>> f) {
+                auto result = f.get();
+                if (result.second) {
+                    return result.first;
+                }
+                for (;result.second;result = do_try_acquire_once().get());
```

Comment:
I guess the condition should be `!result.second`

Reply:
good catch. corrected.